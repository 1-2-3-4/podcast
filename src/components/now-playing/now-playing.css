.now-playing-container {
  padding: 25px;
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
      -ms-flex-flow: column;
          flex-flow: column;
  -webkit-box-align: center;
      -ms-flex-align: center;
          align-items: center;
}

.now-playing-image-container {
  height: 200px;
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-align: center;
      -ms-flex-align: center;
          align-items: center;
}

.now-playing-image {
  width: 200px;
}

.now-playing-title-container {
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-align: center;
      -ms-flex-align: center;
          align-items: center;
  min-height: 38px;
  margin: 20px 0;
  max-width: 300px;
}

.now-playing-title {
  color: #f2f2f2;
  font-weight: 700;
  text-shadow: 1px 1px 10px #292929;
  text-align: center;
}

.now-playing-audio-container {
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-align: center;
      -ms-flex-align: center;
          align-items: center;
}

.prev-button,
.next-button {
  height: 35px;
  width: 35px;
  border: 0;
  border-radius: 100%;
  -webkit-appearance: none;
     -moz-appearance: none;
          appearance: none;
  background-color: #f2f2f2;
  -webkit-box-shadow: inset -2px -2px 6px #c2c2c2, inset 2px 2px 6px #c2c2c2;
          box-shadow: inset -2px -2px 6px #c2c2c2, inset 2px 2px 6px #c2c2c2;
}

.prev-button:hover,
.next-button:hover {
  cursor: pointer;
  background-color: #d6d6d6;
}

.prev-button {
  margin-right: 10px;
}

.next-button {
  margin-left: 10px;
}