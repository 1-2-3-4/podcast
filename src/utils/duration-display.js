export const durationDisplay = (duration) => {
  const minutes = Math.floor((duration / (1000 * 60)) % 60)
  const seconds = Math.floor((duration / 1000) % 60)

  return `${minutes > 10 ? minutes : `0${minutes}`}:${seconds}`
}