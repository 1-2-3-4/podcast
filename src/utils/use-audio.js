import * as React from 'react'
import { AudioContext } from '../components/audio-provider/audio-context'

export const useAudio = () => React.useContext(AudioContext)