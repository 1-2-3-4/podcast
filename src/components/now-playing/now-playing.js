import * as React from 'react'

import { useAudio } from '../../utils/use-audio'

import skipNextIcon from './skip_next.svg'
import skipPrevIcon from './skip_previous.svg'
import './now-playing.css'

export const NowPlaying = () => {
  const { tracks, nowPlayingIndex, setNowPlayingIndex } = useAudio()
  const nowPlayingTrack = tracks[nowPlayingIndex]
  const {
    imageUrl: nowPlayingImageUrl,
    mediaUrl: nowPlayingMediaUrl,
    title: nowPlayingTitle
  } = nowPlayingTrack

  const nextSong = () => {
    const nextIndex = nowPlayingIndex === tracks.length - 1 ? 0 : nowPlayingIndex + 1
    setNowPlayingIndex(nextIndex)
  }

  const prevSong = () => {
    const prevIndex = nowPlayingIndex === 0 ? tracks.length - 1 : nowPlayingIndex - 1
    setNowPlayingIndex(prevIndex)
  }

  return (
    <div className="now-playing-container">
      <div className="now-playing-image-container">
        <img className="now-playing-image" src={nowPlayingImageUrl} alt={nowPlayingTitle}/>
      </div>

      <div className="now-playing-title-container">
        <div className="now-playing-title">{nowPlayingTitle}</div>
      </div>
      
      <div className="now-playing-audio-container">
        <button className="prev-button" onClick={prevSong}>
          <img src={skipPrevIcon} alt="Previous Song" />
        </button>

        <audio
          autoPlay
          controls
          src={nowPlayingMediaUrl}
          onEnded={nextSong}
        >
          {tracks.map(track => {
            const { mediaUrl } = track
            return (
              <source key={mediaUrl} src={mediaUrl} type="audio/mp4" />
            )
          })}
        </audio>

        <button className="next-button" onClick={nextSong}>
          <img src={skipNextIcon} alt="Next Song" />
        </button>
      </div>
    </div>
  )
}