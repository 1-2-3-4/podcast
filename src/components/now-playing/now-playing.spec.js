import * as React from 'react'
import { shallow } from 'enzyme'

import { NowPlaying } from './now-playing'
import * as useAudioUtil from '../../utils/use-audio'

describe('NowPlaying', () => {
  describe('when it renders', () => {
    let wrapper
    const mockTracks = [
      {
        title: 'This is a Fake Song',
        imageUrl: 'https://image.com/this-is-a-fake-song',
        mediaUrl: 'https://media.com/this-is-a-fake-song',
        duration: 10000
      },
      {
        title: 'Also a Mock Song',
        imageUrl: 'https://image.com/also-a-mock-song',
        mediaUrl: 'https://media.com/also-a-mock-song',
        duration: 10000
      },
      {
        title: 'Yup, Still More Fake Songs',
        imageUrl: 'https://image.com/yup-still-more-fake-songs',
        mediaUrl: 'https://media.com/yup-still-more-fake-songs',
        duration: 10000
      }
    ]
    const mockNowPlayingIndex = 2
    const mockNextTrack = 0
    const mockPrevTrack = 1
    const mockSetNowPlayingIndex = jest.fn()

    beforeAll(() => {
      jest.spyOn(useAudioUtil, 'useAudio').mockReturnValue({
        tracks: mockTracks,
        nowPlayingIndex: mockNowPlayingIndex,
        setNowPlayingIndex: mockSetNowPlayingIndex
      })

      wrapper = shallow(<NowPlaying/>)
    })

    afterAll(() => {
      jest.restoreAllMocks()
    })

    it('renders an audio source for each track', () => {
      expect(wrapper.find('source')).toHaveLength(mockTracks.length)
    })

    it('features currently playing track', () => {
      expect(
        wrapper.find({ className: 'now-playing-title' }).contains(mockTracks[mockNowPlayingIndex].title)
      ).toBe(true)
    })

    it('clicking next song button sets the next track', () => {
      wrapper.find({ className: 'next-button' }).simulate('click')
      expect(mockSetNowPlayingIndex).toHaveBeenCalledWith(mockNextTrack)
    })

    it('clicking prev song button sets the prev track', () => {
      wrapper.find({ className: 'prev-button' }).simulate('click')
      expect(mockSetNowPlayingIndex).toHaveBeenCalledWith(mockPrevTrack)
    })

    it('matches snapshot', () => {
      expect(wrapper.getElement()).toMatchSnapshot()
    })
  })
})
