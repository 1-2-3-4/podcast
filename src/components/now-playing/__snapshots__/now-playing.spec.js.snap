// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NowPlaying when it renders matches snapshot 1`] = `
<div
  className="now-playing-container"
>
  <div
    className="now-playing-image-container"
  >
    <img
      alt="Yup, Still More Fake Songs"
      className="now-playing-image"
      src="https://image.com/yup-still-more-fake-songs"
    />
  </div>
  <div
    className="now-playing-title-container"
  >
    <div
      className="now-playing-title"
    >
      Yup, Still More Fake Songs
    </div>
  </div>
  <div
    className="now-playing-audio-container"
  >
    <button
      className="prev-button"
      onClick={[Function]}
    >
      <img
        alt="Previous Song"
        src="skip_previous.svg"
      />
    </button>
    <audio
      autoPlay={true}
      controls={true}
      onEnded={[Function]}
      src="https://media.com/yup-still-more-fake-songs"
    >
      <source
        src="https://media.com/this-is-a-fake-song"
        type="audio/mp4"
      />
      <source
        src="https://media.com/also-a-mock-song"
        type="audio/mp4"
      />
      <source
        src="https://media.com/yup-still-more-fake-songs"
        type="audio/mp4"
      />
    </audio>
    <button
      className="next-button"
      onClick={[Function]}
    >
      <img
        alt="Next Song"
        src="skip_next.svg"
      />
    </button>
  </div>
</div>
`;
